File: generated_code/GPT4/Python/002_SplayTree.py
class SplayTree:
    class Node:
        __slots__ = ('key', 'left', 'right', 'parent')
        def __init__(self, key, parent=None):
            self.key = key
            self.left = None
            self.right = None
            self.parent = parent

    def __init__(self):
        self.root = None

    # Right rotate
    def _right_rotate(self, x):
        y = x.left
        if y is None:
            return
        x.left = y.right
        if y.right:
            y.right.parent = x
        y.parent = x.parent
        if x.parent is None:
            self.root = y
        elif x == x.parent.right:
            x.parent.right = y
        else:
            x.parent.left = y
        y.right = x
        x.parent = y

    # Left rotate
    def _left_rotate(self, x):
        y = x.right
        if y is None:
            return
        x.right = y.left
        if y.left:
            y.left.parent = x
        y.parent = x.parent
        if x.parent is None:
            self.root = y
        elif x == x.parent.left:
            x.parent.left = y
        else:
            x.parent.right = y
        y.left = x
        x.parent = y

    # Splaying operation: brings x to the root
    def _splay(self, x):
        if x is None:
            return
        while x.parent:
            if x.parent.parent is None:
                # Zig step
                if x == x.parent.left:
                    self._right_rotate(x.parent)
                else:
                    self._left_rotate(x.parent)
            elif x == x.parent.left and x.parent == x.parent.parent.left:
                # Zig-Zig step
                self._right_rotate(x.parent.parent)
                self._right_rotate(x.parent)
            elif x == x.parent.right and x.parent == x.parent.parent.right:
                # Zig-Zig step
                self._left_rotate(x.parent.parent)
                self._left_rotate(x.parent)
            elif x == x.parent.right and x.parent == x.parent.parent.left:
                # Zig-Zag step
                self._left_rotate(x.parent)
                self._right_rotate(x.parent)
            elif x == x.parent.left and x.parent == x.parent.parent.right:
                # Zig-Zag step
                self._right_rotate(x.parent)
                self._left_rotate(x.parent)

    # Internal search: returns (node, last_accessed_parent)
    def _find(self, key):
        node = self.root
        prev = None
        while node:
            prev = node
            if key == node.key:
                return node, prev
            elif key < node.key:
                node = node.left
            else:
                node = node.right
        return None, prev

    # Public search: splay the node or its parent to root
    def search(self, key):
        node, last = self._find(key)
        if node:
            self._splay(node)
            return True
        elif last:
            self._splay(last)
        return False

    def insert(self, key):
        if self.root is None:
            self.root = self.Node(key)
            return
        node = self.root
        parent = None
        while node:
            parent = node
            if key == node.key:
                self._splay(node)
                return  # No duplicates in set
            elif key < node.key:
                node = node.left
            else:
                node = node.right
        new_node = self.Node(key, parent)
        if key < parent.key:
            parent.left = new_node
        else:
            parent.right = new_node
        self._splay(new_node)

    def delete(self, key):
        found, _ = self._find(key)
        if not found:
            if _:
                self._splay(_)
            return  # Key not found
        self._splay(found)
        # Now found is at root
        if found.left:
            left_subtree = found.left
            left_subtree.parent = None
            # Find max in left subtree
            max_left = left_subtree
            while max_left.right:
                max_left = max_left.right
            self._splay(max_left)
            # Now max_left is new root, attach right subtree
            max_left.right = found.right
            if found.right:
                found.right.parent = max_left
            self.root = max_left
        elif found.right:
            found.right.parent = None
            self.root = found.right
        else:
            self.root = None
        # Help GC
        found.left = found.right = found.parent = None

    # For debugging: inorder traversal
    def _inorder(self, node, res):
        if node:
            self._inorder(node.left, res)
            res.append(node.key)
            self._inorder(node.right, res)
    def keys(self):
        res = []
        self._inorder(self.root, res)
        return res

    # For checking containment
    def __contains__(self, key):
        return self.search(key)

    # For adding with set-like syntax
    def add(self, key):
        self.insert(key)

    # For removing with set-like syntax
    def remove(self, key):
        self.delete(key)

File: generated_code/GPT4/Java/002_SplayTree.java
public class SplayTree {
    private class Node {
        int key;
        Node left, right, parent;

        Node(int key) {
            this.key = key;
        }
    }

    private Node root;

    // Public insert method
    public void insert(int key) {
        if (root == null) {
            root = new Node(key);
            return;
        }

        Node curr = root, parent = null;
        while (curr != null) {
            parent = curr;
            if (key < curr.key)
                curr = curr.left;
            else if (key > curr.key)
                curr = curr.right;
            else {
                splay(curr);
                return; // Key already exists, do nothing
            }
        }

        Node newNode = new Node(key);
        newNode.parent = parent;
        if (key < parent.key)
            parent.left = newNode;
        else
            parent.right = newNode;
        splay(newNode);
    }

    // Public delete method
    public void delete(int key) {
        Node node = find(key);
        if (node == null)
            return;

        splay(node);

        if (node.left == null) {
            transplant(node, node.right);
        } else if (node.right == null) {
            transplant(node, node.left);
        } else {
            Node min = subtreeMin(node.right);
            if (min.parent != node) {
                transplant(min, min.right);
                min.right = node.right;
                if (min.right != null)
                    min.right.parent = min;
            }
            transplant(node, min);
            min.left = node.left;
            if (min.left != null)
                min.left.parent = min;
        }
        node.left = node.right = node.parent = null; // Help GC
    }

    // Public search method, returns Integer key or null if not found
    public Integer search(int key) {
        Node node = find(key);
        if (node != null) {
            splay(node);
            return node.key;
        }
        return null;
    }

    // Helper: Find node with given key
    private Node find(int key) {
        Node curr = root;
        while (curr != null) {
            if (key < curr.key)
                curr = curr.left;
            else if (key > curr.key)
                curr = curr.right;
            else
                return curr;
        }
        return null;
    }

    // Helper: Splay operation
    private void splay(Node x) {
        if (x == null)
            return;
        while (x.parent != null) {
            if (x.parent.parent == null) {
                // Zig step
                if (x == x.parent.left)
                    rotateRight(x.parent);
                else
                    rotateLeft(x.parent);
            } else if (x == x.parent.left && x.parent == x.parent.parent.left) {
                // Zig-zig step
                rotateRight(x.parent.parent);
                rotateRight(x.parent);
            } else if (x == x.parent.right && x.parent == x.parent.parent.right) {
                // Zig-zig step
                rotateLeft(x.parent.parent);
                rotateLeft(x.parent);
            } else if (x == x.parent.right && x.parent == x.parent.parent.left) {
                // Zig-zag step
                rotateLeft(x.parent);
                rotateRight(x.parent);
            } else {
                // Zig-zag step
                rotateRight(x.parent);
                rotateLeft(x.parent);
            }
        }
        root = x;
    }

    // Helper: Rotate left at node x
    private void rotateLeft(Node x) {
        Node y = x.right;
        if (y == null)
            return;
        x.right = y.left;
        if (y.left != null)
            y.left.parent = x;
        y.parent = x.parent;
        if (x.parent == null)
            root = y;
        else if (x == x.parent.left)
            x.parent.left = y;
        else
            x.parent.right = y;
        y.left = x;
        x.parent = y;
    }

    // Helper: Rotate right at node x
    private void rotateRight(Node x) {
        Node y = x.left;
        if (y == null)
            return;
        x.left = y.right;
        if (y.right != null)
            y.right.parent = x;
        y.parent = x.parent;
        if (x.parent == null)
            root = y;
        else if (x == x.parent.left)
            x.parent.left = y;
        else
            x.parent.right = y;
        y.right = x;
        x.parent = y;
    }

    // Helper: Replace subtree u with subtree v
    private void transplant(Node u, Node v) {
        if (u.parent == null)
            root = v;
        else if (u == u.parent.left)
            u.parent.left = v;
        else
            u.parent.right = v;
        if (v != null)
            v.parent = u.parent;
    }

    // Helper: Find minimum node in subtree
    private Node subtreeMin(Node x) {
        while (x.left != null)
            x = x.left;
        return x;
    }

    // Optional: For testing purposes
    /*
    public void printInOrder() {
        printInOrder(root);
        System.out.println();
    }

    private void printInOrder(Node node) {
        if (node == null) return;
        printInOrder(node.left);
        System.out.print(node.key + " ");
        printInOrder(node.right);
    }
    */
}